### PR TITLE
add semantic search via pgvector with injected EmbeddingService

### PR DESCRIPTION
**Pull request by aaryan-iiitu**
## Work done 

This PR implements semantic search for `PostgresMemoryStore` using `pgvector` which was mentioned in #526 

## Changes made 

- Extracted shared pgvector serialization logic into `PostgresVectorHelpers`
- Added an overloaded `search` method that accepts an `EmbeddingService` via method injection
- Kept the default `search` method failing gracefully when no `EmbeddingService` is available
- Implemented cosine similarity search using pgvector (`<=>`) with score normalization
- Added guards for empty embeddings and consistent error handling
- Extended integration tests using `MockEmbeddingService`

### CODE REVIEW

File: modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala [hunk 1]
```diff
@@ -140,10 +137,66 @@ final class PostgresMemoryStore private[memory] (
     Left(
       ProcessingError(
         "postgres-memory-store",
-        "Semantic search is not yet implemented for PostgresMemoryStore. Requires EmbeddingService integration."
+        "Semantic search is not available without an EmbeddingService. Use search(query, topK, filter, embeddingService)."
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/agent/memory/PostgresMemoryStore.scala]: This keeps the trait-level search method unusable for semantic search and only enables it via PostgresMemoryStore concrete type.
Principle: keep interface behavior consistent and avoid forcing callers to downcast for core capability.
Please inject optional EmbeddingService into store config and make override search() perform semantic search when available, with deterministic fallback when unavailable.

File: modules/core/src/main/scala/org/llm4s/agent/memory/PostgresVectorHelpers.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  def embeddingToString(embedding: Array[Float]): String =
+    embedding.mkString("[", ",", "]")
+
+  def stringToEmbedding(s: String): Array[Float] =
+    if (s == null || s.isEmpty) Array.empty
+    else {
+      val cleaned = s.stripPrefix("[").stripSuffix("]")
+      if (cleaned.isEmpty) Array.empty
+      else Try(cleaned.split(",").map(_.trim.toFloat)).getOrElse(Array.empty)
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/agent/memory/PostgresVectorHelpers.scala]: Parsing failure is silently converted to empty embedding.
Principle: errors should be explicit with context, not silently swallowed.
Please return a typed error from parsing and preserve cause so bad stored vectors are diagnosable.

File: modules/core/src/test/scala/org/llm4s/agent/memory/PostgresMemoryStoreSpec.scala [hunk 3]
```diff
@@ -82,4 +82,41 @@ class PostgresMemoryStoreSpec extends AnyFlatSpec with Matchers with BeforeAndAf
     result.toOption.flatten.map(_.content) shouldBe Some("Persistence Check")
     store2.close()
   }
+
+  it should "perform semantic search using overloaded method with EmbeddingService" in skipIfDisabled {
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/agent/memory/PostgresMemoryStoreSpec.scala]: Only happy path is covered here, which matches the codecov patch failure.
Principle: new behavior must include failure-path and edge-case tests.
Please add tests for empty query embedding, embedding service failure, and default search() error path.

**Review by rorygraves — CHANGES_REQUESTED** [review]
## Re-Review (v3): REQUEST CHANGES

Good progress — all four original review items have been addressed:

1. ✅ **Constructor injection** — `PostgresMemoryStore` now accepts `Option[EmbeddingService]` with graceful fallback via `recall()`
2. ✅ **Silent error swallowing** — `PostgresVectorHelpers.stringToEmbedding` returns `Result[Array[Float]]` with error context
3. ✅ **Helper extraction** — Both stores now use shared `PostgresVectorHelpers`
4. ✅ **Test coverage** — Tests added for semantic search, fallback, empty vectors, service failures, score clamping

### Remaining blocking issues

#### 1. Merge conflicts
The PR is currently in `CONFLICTING` state. Please rebase onto latest `main` and resolve conflicts.

#### 2. Cross-package dependency direction
`PgVectorStore` (in `org.llm4s.vectorstore`) now imports from `org.llm4s.agent.memory.PostgresVectorHelpers`. This creates an undesirable dependency from the general-purpose `vectorstore` package into the `agent.memory` package.

**Fix:** Move `PostgresVectorHelpers` to `org.llm4s.vectorstore` (or a shared utility package), then import from there in both stores.

#### 3. No full CI build visible
Only `comment-on-external-pr` passes. After resolving conflicts, please ensure the full build/test suite passes.

### Non-blocking suggestions
- Consider removing the delegating wrapper methods in the `PostgresMemoryStore` companion (`embeddingToString`/`stringToEmbedding`) since callers can use `PostgresVectorHelpers` directly
- The repeated while-loop + mutable error pattern could be extracted into a `collectResults[T]` helper (can be a follow-up)
- First test reduces assertion specificity (only checks `shouldBe defined` instead of verifying content/metadata round-trip)

The core implementation is solid — once conflicts are resolved and the helper is relocated, this should be ready to merge.

**Comment by aaryan-iiitu** [comment]
@rorygraves  I rebased onto latest main, resolved merge conflicts, and moved PostgresVectorHelpers to org.llm4s.vectorstore to fix the cross package dependency direction. Kindly take a look at it whenever you have time.

File: modules/core/src/test/scala/org/llm4s/agent/memory/PostgresMemoryStoreSpec.scala [hunk 4]
```diff
@@ -82,4 +78,79 @@ class PostgresMemoryStoreSpec extends AnyFlatSpec with Matchers with BeforeAndAf
...
+      "I like apples",
+      MemoryType.Task,
+      embedding = Some(embedding)
+    )
+    store.store(relevant)
+    store
+      .search("apple", 1, MemoryFilter.All)
+      .fold(
+        e => fail(s"Search failed: ${e.message}"),
+        results => results.head.memory.content shouldBe "I like apples"
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/agent/memory/PostgresMemoryStoreSpec.scala]: `results.head` is still partial and can fail with an unhelpful exception when search returns empty. Assert nonEmpty first, then inspect first element via safe pattern. Principle: tests should fail with clear intent, not collection access errors.

File: modules/core/src/test/scala/org/llm4s/agent/memory/PostgresMemoryStoreSpec.scala [hunk 5]
```diff
@@ -82,4 +78,79 @@ class PostgresMemoryStoreSpec extends AnyFlatSpec with Matchers with BeforeAndAf
...
+        memories => {
+          memories.nonEmpty shouldBe true
+          memories.head.score shouldBe 0.0
+        }
+      )
+    storeNoEmb.close()
+  }
+
+  it should "clamp similarity scores to [0, 1]" in skipIfDisabled {
+    store.search("anything", 5, MemoryFilter.All).foreach { results =>
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/agent/memory/PostgresMemoryStoreSpec.scala]: This test can pass vacuously when `results` is empty, so score clamping is not actually validated. Seed one vector row (or assert nonEmpty) before checking bounds. Principle: assertions should prove behavior, not just execute code paths.

File: modules/core/src/test/scala/org/llm4s/agent/memory/PostgresMemoryStoreSpec.scala [hunk 6]
```diff
@@ -62,7 +59,6 @@ class PostgresMemoryStoreSpec extends AnyFlatSpec with Matchers with BeforeAndAf
     )
 
     store.store(memory).isRight shouldBe true
-
     val retrieved = store.get(id).toOption.flatten
     retrieved shouldBe defined
     retrieved.get.content shouldBe "Hello, I am a test memory"
```
Comment (vim89) [inline modules/core/src/test/scala/org/llm4s/agent/memory/PostgresMemoryStoreSpec.scala]: Using `.get` here keeps a runtime throw path in a new test. Prefer `fold`/pattern-match and assert the `Some` branch explicitly. Principle: avoid partial access on optional values and keep tests total/deterministic.
